metrics/src/handler.rs

```diff
@@ -0,0 +1,144 @@
+use crate::data::Metrics;
+use ibc::events::IbcEvent;
+use ibc_proto::google::protobuf::Any;
+use std::cell::Cell;
+use std::time::{Duration, Instant};
+
+pub struct MetricsHandler {
+	metrics: Metrics,
+	last_send_packet_event_time: Cell<Option<Instant>>,
+	last_receive_packet_event_time: Cell<Option<Instant>>,
+	last_acknowledge_packet_event_time: Cell<Option<Instant>>,
+	last_sent_packet_time: Cell<Option<Instant>>,
+	last_sent_acknowledgment_time: Cell<Option<Instant>>,
+	last_sent_timeout_packet_time: Cell<Option<Instant>>,
+}
+
+impl MetricsHandler {
+	pub fn new(metrics: Metrics) -> Self {
+		Self {
+			metrics,
+			last_send_packet_event_time: Cell::new(None),
+			last_receive_packet_event_time: Cell::new(None),
+			last_acknowledge_packet_event_time: Cell::new(None),
+			last_sent_packet_time: Cell::new(None),
+			last_sent_acknowledgment_time: Cell::new(None),
+			last_sent_timeout_packet_time: Cell::new(None),
+		}
+	}
+
+	pub async fn handle_events(&self, events: &[IbcEvent]) {
+		for event in events {
+			match event {
+				IbcEvent::SendPacket(_) => {
+					self.metrics.number_of_send_packet_events.inc();
+					if let Some(elapsed) =
+						get_and_update_delta_time(&self.last_send_packet_event_time)
+					{
+						self.metrics.send_packet_event_time.observe(elapsed.as_millis() as f64);
+					}
+				},
+				IbcEvent::ReceivePacket(_) => {
+					self.metrics.number_of_undelivered_packets.dec();
+					self.metrics.number_of_receive_packet_events.inc();
+					if let Some(elapsed) =
+						get_and_update_delta_time(&self.last_receive_packet_event_time)
+					{
+						self.metrics.receive_packet_event_time.observe(elapsed.as_millis() as f64);
+					}
+				},
+				IbcEvent::AcknowledgePacket(_) => {
+					self.metrics.number_of_undelivered_acknowledgements.dec();
```

Comment:
Let this metric be number of received acknowledgements instead
```suggestion
					self.metrics.number_of_undelivered_acknowledgements.dec();
```